Update Typescript migration documentation for Merging (#2134)
MarkBind/markbind: docs/devGuide/development/migratingToTypeScript.md

You are now ready to create a pull request for the changes to the repository.

### (Optional) Step 4: Changing the commit

While refactoring you may encounter scenarios where you might have to amend your code after a commit. There are 2 common scenarios which we will cover. 

> ##### Scenario 1 - you have 2 commits and want to add on to the second commit: 

Reviewer: nit: can we exclude these level-5 headers from the page nav? The headers are quite long and may clutter up the page nav.
Or maybe we can put it in a panel since this section is optional?

Author: Good idea, ill put this into a panel!

To append a new commit onto the already existing commit you can do the following: 
1. Make the changes you want to add to the second commit
2. Stage the changes using `git add`
3. Run `git commit --amend`
4. This will open your default text editor, where you can edit the commit message for the second commit to reflect the new changes you have made
5. Save and close the text editor. This will replace the second commit with the updated changes.

<box type="warning">

> It's important to note that amending a commit changes its commit hash, which means that if you've already pushed the original commit to a remote repository, you'll need to force push your changes.

Reviewer: ```suggestion
Amending a commit changes its commit hash, which means that if you've already pushed the original commit to a remote repository, you'll need to force push your changes.
```

</box>

> ##### Scenario 2 - you have 3 commits and want to make it into 2 commits:

Reviewer: ```suggestion
##### Scenario 2 - you have >2 commits and want to make it into 2 commits:
```

> 1. Run `git rebase -i HEAD~3` to start an interactive rebase session. This will show a list of the three most recent commits in your text editor.

Reviewer: ```suggestion
1. Run `git rebase -i master` to start an interactive rebase session. This will show a list of the recent commits in your text editor.
```

Rebasing on master can allow this command to be used regardless of how many commits there are! Also bonus of making the rebase commit (for merging into master) easier later on.

The solution to this is to make the rename and adapt commits separate, in order for Git to recognize first that the file has changed into `.ts`, and so the changes are compared not against the old `.js` file, but to the renamed `.ts` file instead.

### Why do we need a rebase and merge commit

Even if the migration developer has kept the history intact through the separate "Rename" and "Adapt" commits, this is only intact at their working branch. At the end of the day, the totality of the changes in the working branch is compared against the target branch, in which the same consideration would apply.

If we do the usual squash commit, the changes from the two commits are combined into a new commit and only that commit will be pushed into the target branch. The original two commits are omitted, therefore the history of the working branch that we have tried to keep intact is stripped away.

> The reason we use a rebase operation is because it updates the base branch of a

Reviewer 2: I think the last sentence is fine. I think there's no need to define what a rebase is.

The solution to this is to make the rename and adapt commits separate, in order for Git to recognize first that the file has changed into `.ts`, and so the changes are compared not against the old `.js` file, but to the renamed `.ts` file instead.

### Why do we need a rebase and merge commit

Even if the migration developer has kept the history intact through the separate "Rename" and "Adapt" commits, this is only intact at their working branch. At the end of the day, the totality of the changes in the working branch is compared against the target branch, in which the same consideration would apply.

> If we do the usual squash commit, the changes from the two commits are combined into a new commit and only that commit will be pushed into the target branch. The original two commits are omitted, therefore the history of the working branch that we have tried to keep intact is stripped away. This results in a linear history, as if the changes in the working branch were made

Reviewer: I don't think the point about linear history is necessarily relevant here - our repo history is quite linear because of the squash and merge strategy so using a rebase over squash+merge doesn't add much linearity. Maybe can we revert this change?

Reviewer 3: @ong6 this is not reverted yet?
